feat: add tenantID template function

FUSAKLA opened the pull request:
As discussed in this issue https://github.com/grafana/mimir/issues/3121 and as a followup from https://github.com/prometheus/alertmanager/pull/3174 using the additional options.

I suppose we will wait for some AM release, not to import random commits?

Ad tests, I wanted to do some more on point test, but it's hard to make the AM to just render a template :/ 
So at least testing the option.  

closes  #3121

FUSAKLA commented:
@gotjosh PTAL if you find a while since you have the context :pray:  (no rush just so it does not stay forgotten)

pracucci commented on go.mod:
```diff
@@ -18,7 +18,7 @@ require (
 	github.com/gorilla/mux v1.8.0
 	github.com/grafana/dskit v0.0.0-20221216094413-a9826f9b0468
 	github.com/grafana/e2e v0.1.1-0.20221115065638-fe4609fcbc71
-	github.com/hashicorp/golang-lru v0.5.4
```
There are many vendor updates here. Are all of these required?

FUSAKLA replied:
Unfortunately, it looks like it, or we would need to tweak it in the go.mod using some replace?
I suppose that was caused by [the upgrade of the golang-lru in alertmanager](https://github.com/prometheus/alertmanager/commit/c2b2defe48e7afe1db59c88da5643d0c920634d2) :/

pracucci commented on pkg/alertmanager/alertmanager.go:
```diff
@@ -300,6 +302,16 @@ func clusterWait(position func() int, timeout time.Duration) func() time.Duratio
+	funcs := tmpltext.FuncMap{"tenantID": func() string { return userID }}
+	return func(text *tmpltext.Template, html *tmplhtml.Template) {
+		text.Funcs(funcs)
+		html.Funcs(tmplhtml.FuncMap(funcs))
```
[nit] Should't be required here.

```suggestion
		html.Funcs(funcs)
```

FUSAKLA replied:
Thanks, fixed in https://github.com/grafana/mimir/pull/3758/commits/975295f6c74afad4ef5d2d47b8f68ba8fd34f48e

pracucci commented on pkg/alertmanager/api.go:
```diff
@@ -256,7 +256,7 @@ func validateUserConfig(logger log.Logger, cfg alertspb.AlertConfigDesc, limits
 	}
 
-	_, err = template.FromGlobs(templateFiles...)
+	_, err = template.FromGlobs(templateFiles)
```
I think we need to inject the tenant ID function there as well, otherwise a template using the function may fail.  Can you verify it adding a unit test too, please? You can add a test case to `TestAMConfigValidationAPI`.

FUSAKLA replied:
Thanks, fixed in https://github.com/grafana/mimir/pull/3758/commits/975295f6c74afad4ef5d2d47b8f68ba8fd34f48e

pracucci reviewed (commented):
Thanks for working on this! Overall LGTM. I left a couple of minor comments. Can you also add a CHANGELOG entry and rebase please?

The vendored alertmanager update adds Webex support. It should be also mentioned in the CHANGELOG.

pracucci commented:
@FUSAKLA I pushed a commit to add Webex support to Mimir. It was easier to do than explaining it 😉 I hope you don't mind!
https://github.com/grafana/mimir/pull/3758/commits/41c55ad22b25a982dda009747262126b1495ad78

FUSAKLA commented:
@pracucci not at all, thanks for that! 

Sorry about the delay, needed to take a break over holidays :christmas_tree: 

Should be rebased now, added the changelog entry and resolved all the comments, thanks for looking into it.
There is still the question of the vendor updates, which I'm not sure how to deal with. 

I also added a note about the templating to the docs in https://github.com/grafana/mimir/pull/3758/commits/dceb21cac964069c219f6dee7ebab5fedafebe91
![image](https://user-images.githubusercontent.com/6112562/210380081-ef0a9ba7-bdba-454d-a3f2-43d25bba582a.png)

PTAL

pracucci reviewed (approved):
Thanks @FUSAKLA, LGTM! I just pushed a nit change to CHANGELOG and a smoke test in the config validation API. I hope you don't mind 🤗

FUSAKLA commented:
Great, thanks!

FUSAKLA commented:
No worries @gotjosh, thank you for guiding the AM work. That made it way easier :)
